csrc/includes/cpu_adam.h

```diff
@@ -169,23 +198,45 @@ class Adam_Optimizer {
 
 #if defined(__AVX512__) or defined(__AVX256__)
 template <int span>
-void Adam_Optimizer::Step_AVX(size_t* rounded_size,
+std::pair<float, float> Adam_Optimizer::Step_AVX(size_t* rounded_size,
                               float* _params,
                               float* grads,
                               float* _exp_avg,
                               float* _exp_avg_sq,
                               size_t _param_size,
                               ds_half_precision_t* dev_params,
-                              bool half_precision)
+                              bool half_precision_param,
+                              bool half_precision_opt_state,
+                              float avg_scaling, 
+                              float avg_sq_scaling, 
+                              float new_avg_scaling, 
+                              float new_avg_sq_scaling)
 {
     size_t new_rounded_size = 0;
-    int rshft = half_precision ? 1 : 0;
+    int rshft_param = half_precision_param ? 1 : 0;
+    int rshft_opt_state = half_precision_opt_state ? 1 : 0;
 
     AVX_Data betta1_4;
     betta1_4.data = SIMD_SET(_betta1);
     AVX_Data betta2_4;
     betta2_4.data = SIMD_SET(_betta2);
 
+
+    AVX_Data avg_scaling_4;
+    avg_scaling_4.data = SIMD_SET(avg_scaling);
+    AVX_Data avg_sq_scaling_4;
+    avg_sq_scaling_4.data = SIMD_SET(avg_sq_scaling);
+
+    AVX_Data new_avg_scaling_4;
+    new_avg_scaling_4.data = SIMD_SET(new_avg_scaling);
+    AVX_Data new_avg_sq_scaling_4;
+    new_avg_sq_scaling_4.data = SIMD_SET(new_avg_sq_scaling);
+
+    int OMP_MAX_THREADS = omp_get_max_threads();
+    float max_avg[OMP_MAX_THREADS], max_avg_sq[OMP_MAX_THREADS];
+    std::fill(max_avg, max_avg + OMP_MAX_THREADS, 0.0f);
+    std::fill(max_avg_sq, max_avg_sq + OMP_MAX_THREADS, 0.0f);
```

Comment:
if I understand this part is related to Z3 and the serves as an initialization step, can we maybe move them to a different function and call it conditionally?

Reply:
Let me think a little bit, it should be doable.